virtio-devices/src/balloon.rs

```diff
@@ -198,9 +280,12 @@ impl BalloonEpollHandler {
 
                 match queue_index {
                     0 => {
-                        Self::release_memory_range(desc_chain.memory(), range_base, range_len)?;
+                        self.release_memory_range(desc_chain.memory(), range_base, range_len)?;
                     }
                     1 => {
+                        // SAFETY: FFI call. Trivially safe.
+                        let page_size: u64 = unsafe { sysconf(_SC_PAGESIZE) as u64 };
+                        let range_base = range_base & !(page_size - 1);
```

Comment:
I think it is better to have a separate commit for the changes to the deflation with well documentation about its behavior.
Also, what's the behavior of "madvise" when the `range_len` is smaller than the system page size?

Reply:
> I think it is better to have a separate commit for the changes to the deflation with well documentation about its behavior.

Should we do the change for deflation? I see there is nothing changed in qemu. See https://github.com/qemu/qemu/blob/4329d049d5b8d4af71c6b399d64a6d1b98856318/hw/virtio/virtio-balloon.c#L145

> Also, what's the behavior of "madvise" when the `range_len` is smaller than the system page size?

madvise will check if the memory range start page size allgned. If not, error will be thrown out. If so, range_len will be alligned down to page size. That is if range_len is less then page size, no memory will be freed. See https://github.com/torvalds/linux/blob/6aeadf7896bff4ca230702daba8788455e6b866e/mm/madvise.c#L832

Reply:
> > I think it is better to have a separate commit for the changes to the deflation with well documentation about its behavior.
> 
> Should we do the change for deflation? I see there is nothing changed in qemu. See https://github.com/qemu/qemu/blob/4329d049d5b8d4af71c6b399d64a6d1b98856318/hw/virtio/virtio-balloon.c#L145
> 

From the link you provided above, I believe what QEMU does for deflateation is at the granularity of actual page size. We can do the same thing, and it will be a good improvement as it will enable our virtio-ballon to support non-4k pages for deflation.

> > Also, what's the behavior of "madvise" when the `range_len` is smaller than the system page size?
> 
> madvise will check if the memory range start page size allgned. If not, error will be thrown out. If so, range_len will be alligned down to page size. That is if range_len is less then page size, no memory will be freed. See https://github.com/torvalds/linux/blob/6aeadf7896bff4ca230702daba8788455e6b866e/mm/madvise.c#L832

Great to know. Thank you for sharing.

Reply:
> > > I think it is better to have a separate commit for the changes to the deflation with well documentation about its behavior.
> > 
> > 
> > Should we do the change for deflation? I see there is nothing changed in qemu. See https://github.com/qemu/qemu/blob/4329d049d5b8d4af71c6b399d64a6d1b98856318/hw/virtio/virtio-balloon.c#L145
> 
> From the link you provided above, I believe what QEMU does for deflateation is at the granularity of actual page size. We can do the same thing, and it will be a good improvement as it will enable our virtio-ballon to support non-4k pages for deflation.

Yes, I will do the same change for clh.